core/src/main/java/org/wso2/carbon/kernel/internal/configprovider/ConfigProviderImpl.java

```diff
@@ -0,0 +1,241 @@
+                namespace = configuration.namespace();
+            }
+        }
+        // lazy loading deployment.yaml configuration.
+        loadDeploymentConfiguration(configFileReader);
+
+        if (namespace != null && deploymentConfigs.containsKey(namespace)) {
+            String jsonConfigString = deploymentConfigs.get(namespace);
+            String yamlConfigString = ConfigurationUtils.convertJSONToYAML(jsonConfigString);
+            if (logger.isDebugEnabled()) {
+                logger.debug("class name: " + configClass.getSimpleName() + " | new configurations: \n" +
+                        yamlConfigString);
+            }
+            String yamlProcessedString = processPlaceholder(yamlConfigString);
+            yamlProcessedString = org.wso2.carbon.kernel.utils.Utils.substituteVariables(yamlProcessedString);
+            Yaml yaml = new Yaml(new CustomClassLoaderConstructor(configClass, configClass.getClassLoader()));
+            yaml.setBeanAccess(BeanAccess.FIELD);
+            return yaml.loadAs(yamlProcessedString, configClass);
+        } else {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Deployment configuration mapping doesn't exist: " +
+                        "creating configuration instance with default values");
+            }
+            try {
+                return configClass.newInstance();
+            } catch (InstantiationException | IllegalAccessException e) {
+                throw new CarbonConfigurationException("Error while creating configuration instance: "
+                        + configClass.getSimpleName(), e);
+            }
+        }
+    }
+
+    @Override
+    public Map getConfigurationMap(String namespace) throws CarbonConfigurationException {
+        // lazy loading deployment.yaml configuration, if it is not exists
+        loadDeploymentConfiguration(configFileReader);
+        // check for json configuration from deployment configs of namespace.
+        if (deploymentConfigs.containsKey(namespace)) {
+            String jsonConfigString = deploymentConfigs.get(namespace);
+            String jsonProcessedString = processPlaceholder(jsonConfigString);
+            jsonProcessedString = org.wso2.carbon.kernel.utils.Utils.substituteVariables(jsonProcessedString);
+            Yaml yaml = new Yaml();
+            return yaml.loadAs(jsonProcessedString, Map.class);
+        }
+        logger.error("configuration doesn't exist for the namespace: " + namespace + " in deployment yaml. Hence " +
+                "return null object");
+        return null;
+    }
+
+
+    /**
+     * This method loads deployment configs in deployment.yaml. loads only if config
+     */
+    private void loadDeploymentConfiguration(ConfigFileReader configFileReader) throws CarbonConfigurationException {
+        if (deploymentConfigs == null) {
+            synchronized (this) {
+                if (deploymentConfigs == null) {
```

Comment:
this method loads the configuration from the deployment.yaml on request if config map is null. if two or more requests comes at the same time when deploymentConfigs is null (first call), all request will read the file to load deployment configuration. to avoid that I added synchronized block.